Refactor README.md (#65)
grafana/xk6-disruptor: README.md

Don't believe us? Check out the following example!

```js
import { PodDisruptor } from "k6/x/disruptor";

> ⚠️  xk6-disruptor is in the alpha stage, undergoing active development. We do not guarantee API compatibility between releases - your k6 scripts may need to be updated on each release until this extension reaches v1.0 release.
export default function () {
    // Create a new pod disruptor with a selector 
    // that matches pods from the "default" namespace with the label "app=my-app".
    const disruptor = new PodDisruptor({
        namespace: "default",
        select: { labels: { app: "my-app" } },
    });

## Use case for xk6-disruptor
    // Check that there is at least one target.
    const targets = disruptor.targets();
    if (targets.length != 1) {
        throw new Error("expected list to have one target");
    }

> The main use case for xk6-disruptor is to test the resiliency of an application of diverse types disruptions by reproducing their effects, but without having to reproduce their root-causes. For example, inject delays in the HTTP requests an application makes to a service  without having to stress or interfere with the infrastructure (network, nodes) on which the service runs, or affecting other workloads in unexpected ways.

Reviewer: Why are you removing the use case? I think it is important to provide context.

Author: Adding it back! Yup, agree.

    // Disrupt the targets by injecting HTTP faults into them for 30 seconds.
    disruptor.injectHTTPFaults(
        {
            averageDelay: 500,
            errorRate: 0.1,
            errorCode: 500,
        },
        30
    );
}
```

> In this way, xk6-disruptor make reliability tests repeatable, predictable and limits their blast-radius. These are important characteristic in order to incorporate this kind of tests in the test suits of applications deployed on shared infrastructures such as staging environments.

Reviewer: I think the concept in this line is relevant, maybe we can rephrase it to make lighter.

Author: Tried to :eyes: 

Any thoughts?

> The project at this time is highly focused on providing an excellent experience to test systems running in Kubernetes. Other platforms are not supported at this time.

Reviewer: It sounds like a piece of marketing to me. I personally don't feel comfortable using this language in the repository of an open source project.

Reviewer: For me the problematic phrase is "excellent experience".

Author: Yup. I removed it :+1:

</br>
</br>

<p align="center">
  <img src="assets/logo.png" alt="xk6-disruptor" width="300"></a>
  <br>
  "Like Unit Testing, but for <strong>Reliability</strong>"
  <br>
</p>
<p align="center">  
</div>

> <blockquote align="center">
> ⚠️ <strong>Important</strong> ⚠️ 
> <br>This project is still in its early stages and under active development.
> <br>The API is subject to change, and there may be bugs.
> <br>Use at your own risk. Thanks!
> </blockquote>

Reviewer: I don't like how this note looks centered below the logo.  I would move below the next paragraph

```suggestion
```

Reviewer: I don't find how to add this block after the next paragraph because there is already another suggestion on that paragraph. I think you can add this block later in another commit:

Author: Let me try :+1:

> This extension adds fault injection capabilities to [Grafana k6](https://github.com/grafana/k6). It implements the ideas of the Chaos Engineering discipline and enables Grafana k6 users to test their system's reliability under turbulent conditions.

Author: I'm sometimes annoyed when the title equals the start of the first sentence.

But yeah, accepting it.

  - No need to try to glue multiple tools together to get the job done.

All that plus great docs and sane APIs. Also, this project has been built to be a good citizen in the Grafana k6 ecosystem by:

- Working well with other extensions.
- Working well with k6's core concepts and features.

You can check this out in the following example:

```js
import { PodDisruptor } from "k6/x/disruptor";

export default function () {
    // Create a new pod disruptor with a selector 
    // that matches pods from the "default" namespace with the label "app=my-app"
    const disruptor = new PodDisruptor({
        namespace: "default",
        select: { labels: { app: "my-app" } },
    });

    // Check that there is at least one target
    const targets = disruptor.targets();
    if (targets.length != 1) {
        throw new Error("expected list to have one target");
    }

    // Disrupt the targets by injecting HTTP faults into them for 30 seconds
    disruptor.injectHTTPFaults(
        {
            averageDelay: 500,
            errorRate: 0.1,
            errorCode: 500,
        },
        30
    );

> }

Author: I just wanted to compress the script a bit so it doesn't take all the screen state. 

I don't have strong opinions here, either.

> }

Reviewer: Then I would prefer the way I suggest which is consistent with the examples in the documentation.

```

## Features

The project, at this time, is intended to test systems running in Kubernetes. Other platforms are not supported at this time.

> Right now, it offers an [API](https://k6.io/docs/javascript-api/xk6-disruptor/api) for creating disruptors that target one specific type of the component (e.g., Pods) and is capable of injecting different kinds of [faults]([Services](https://k6.io/docs/javascript-api/xk6-disruptor/api/faults), such as errors in HTTP requests served by that component. 

Reviewer: ```suggestion
It offers an [API](https://k6.io/docs/javascript-api/xk6-disruptor/api) for creating disruptors that target one specific type of the component (e.g., Pods) and is capable of injecting different kinds of [faults]([Services](https://k6.io/docs/javascript-api/xk6-disruptor/api/faults), such as errors in HTTP requests served by that component. 
```

Currently, disruptors exist for [Pods](https://k6.io/docs/javascript-api/xk6-disruptor/api/poddisruptor) and [[Services](https://k6.io/docs/javascript-api/xk6-disruptor/api/servicedisruptor), but others will be introduced in the future as well as additional types of faults for the existing disruptors.

## Use cases

The main use case for xk6-disruptor is to test the resiliency of an application of diverse types of disruptions by reproducing their effects without reproducing their root causes. For example, inject delays in the HTTP requests an application makes to a service without having to stress or interfere with the infrastructure (network, nodes) on which the service runs or affect other workloads in unexpected ways.

In this way, xk6-disruptor make reliability tests repeatable and predictable while limiting their blast radius. These are essential characteristics to incorporate these tests in the test suits of applications deployed on shared infrastructures such as staging environments.

## Learn more

> Check the [get started guide](https://k6.io/docs/javascript-api/xk6-disruptor/getstarted) for instructions on how to install and use `xk6-disruptor`.

Reviewer: ```suggestion
Check the [get started guide](https://k6.io/docs/javascript-api/xk6-disruptor/get-started) for instructions on how to install and use `xk6-disruptor`.
```

</br>
</br>

<p align="center">
  <img src="assets/logo.png" alt="xk6-disruptor" width="300"></a>
  <br>
  "Like Unit Testing, but for <strong>Reliability</strong>"
  <br>
</p>
<p align="center">  
</div>

xk6-disruptor is an extension adds fault injection capabilities to [Grafana k6](https://github.com/grafana/k6). It implements the ideas of the Chaos Engineering discipline and enables Grafana k6 users to test their system's reliability under turbulent conditions.

<blockquote>
⚠️ <strong>Important</strong> ⚠️ 

> <br>This project is still in its early stages and under active development.
> <br>The API is subject to change, and there may be bugs.
> <br>Use at your own risk. Thanks!

Reviewer: ```suggestion
> ⚠️  xk6-disruptor is in the alpha stage, undergoing active development. We do not guarantee API compatibility between releases - your k6 scripts may need to be updated on each release until this extension reaches v1.0 release.
```

Reviewer: @dgzlopes In an early version of the README we decided that "use at your own risk" may deter people form using the tool. We prefer this wording.

Author: Sounds good!

## Why another Chaos Engineering tool?

Compared to other tools in the space, this one is purposely designed and built to provide the best experience for developers trying to make their systems more reliable:

- Everything as code.

>   - No need to mess with YAML or learn a new DSL.

Reviewer: I like the "no need to learn a new DSL" but I don't feel comfortable with "mess with YAML" part. I find it a weak differentiation mostly considering we are in the Kubernetes ecosystem on which yaml is still prevalant. 
```suggestion
  - No need to learn a new DSL.
```

- Fast to adopt with no day-two surprises.
  - No need to deploy and maintain a fleet of agents or operators.

> - Easy to extend and integrate with other types of tests.

Reviewer: I like the concept, not sure it is clear enough. Make the assumption the person who is reading is already familiar with the other types of tests that k6 can do. I would be more explicit.

Author: ```suggestion
- Easy to extend and integrate with other [types of tests](https://k6.io/docs/test-types/introduction/).
```

Yeah, agree. What do you think if we link to the docs?

Reviewer: That should work.

>   - No need to try to glue five tools together to get the job done.

Reviewer: ```suggestion
  - No need to try to glue multiple tools together to get the job done.
```

All that plus great docs and sane APIs. Also, this project has been built to be a good citizen in the Grafana k6 ecosystem by:

- Working well with other extensions.
- Working well with k6's core concepts and features.

You can check this out in the following example:

```js
import { PodDisruptor } from "k6/x/disruptor";

export default function () {
    // Create a new pod disruptor with a selector 
    // that matches pods from the "default" namespace with the label "app=my-app"
    const disruptor = new PodDisruptor({
        namespace: "default",
        select: { labels: { app: "my-app" } },
    });

    // Check that there is at least one target
    const targets = disruptor.targets();
    if (targets.length != 1) {
        throw new Error("expected list to have one target");
    }

    // Disrupt the targets by injecting HTTP faults into them for 30 seconds
    disruptor.injectHTTPFaults(
        {
            averageDelay: 500,
            errorRate: 0.1,
            errorCode: 500,
        },
        30
    );
}
```

## Features

> =======

Author: Should we remove this? @pablochacin

Reviewer: I think so

</br>
</br>

<p align="center">
  <img src="assets/logo.png" alt="xk6-disruptor" width="300"></a>
  <br>
  "Like Unit Testing, but for <strong>Reliability</strong>"
  <br>
</p>
<p align="center">  
</div>

xk6-disruptor is an extension adds fault injection capabilities to [Grafana k6](https://github.com/grafana/k6). It implements the ideas of the Chaos Engineering discipline and enables Grafana k6 users to test their system's reliability under turbulent conditions.

> <blockquote align="center">

Author: Here? Or did I misunderstand what you meant? :D @pablochacin

Reviewer: Yes, only that not centered.

> <blockquote align="center">

Reviewer: ```suggestion
<blockquote >
```
